### PR TITLE
Print statistics in cmdline mode as in GUI

### LINKED ISSUE #2639: Output same info from GUI via command line [$50]
As per [forum](http://forum.openscad.org/Get-console-information-in-order-to-store-it-tp25022.html), the output in the console from render:
```
Compiling design (CSG Tree generation)...
ECHO: version = [2018, 11, 4]
Rendering Polygon Mesh using CGAL...
Geometries in cache: 5
Geometry cache size in bytes: 754208
CGAL Polyhedrons in cache: 3
CGAL cache size in bytes: 26485440
Total rendering time: 0 hours, 0 minutes, 21 seconds
   Top level object is a 3D object:
   Simple:        yes
   Vertices:     4276
   Halfedges:   20384
   Edges:       10192
   Halffacets:  11816
   Facets:       5908
   Volumes:         2
Rendering finished.
```
should be able to be collected somehow via command line.

Similarly the preview info would also be useful
```
Compiling design (CSG Tree generation)...
ECHO: version = [2018, 11, 4]
Compiling design (CSG Products generation)...
Geometries in cache: 5
Geometry cache size in bytes: 754208
CGAL Polyhedrons in cache: 3
CGAL cache size in bytes: 26485440
Compiling design (CSG Products normalization)...
Compiling highlights (1 CSG Trees)...
Normalized CSG tree has 4 elements
Compile and preview finished.
Total rendering time: 0 hours, 0 minutes, 0 seconds
```

Note @nophead comment 'The statistics about edges and times, etc, are generated by the GUI code in mainwin.cc, so simply don't exist in command line mode.'

Having the output would also be useful to expand the test suite and for performance comparisons between releases.

Discussion on format;  plain text output or would an encoded format be more useful, presumably JSON?

<bountysource-plugin>

---
There is a **[$50 open bounty](https://www.bountysource.com/issues/67865416-output-same-info-from-gui-via-command-line?utm_campaign=plugin&utm_content=tracker%2F52063&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F52063&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Fat-Zer**
- Added a helper visitor class to uniformly process Geometry
 - Move statistic printing to a dedicated class
 - Resolves #2639

### CODE REVIEW

**Comment by t-paul** [comment]
The AppVeyor failure seems to be just a timeout, kicking off a new build...

**Comment by Fat-Zer** [comment]
By the word, I've forgot to mention on the commit that I've changed the format of time elapsed by rendering from

    Total rendering time: h hours m minutes s seconds

to

    Total rendering time: mm:ss.zzz

I hope it's not a big deal... Even for complicated models the rendering rarely takes up a whole hour, so IMHO the user might be more interested in microseconds rather than hours.

**Comment by MichaelAtOz** [comment]
a. I've had models take over 24 hours
b. the formal should be consistent between GUI & CMD line, so it can be parsed regardless.

If you want milliseconds do it on both.

**Comment by t-paul** [comment]
The code is nicely factored out so same format is not a problem. Maybe there's a resonable way to achieve a more dynamic format, like put in hours only when actually needed?

**Comment by Fat-Zer** [comment]
@MichaelAtOz, are such renders anyway common? Or is it just a simple worst-case example?<br> 
I maybe was wrong in my assumptions, if you wish, I won't hesitate reverting the change...

And sure, it has the same format for GUI&CLI ;)

**Comment by Fat-Zer** [comment]
>The code is nicely factored out so same format is not a problem. Maybe there's a resonable way to achieve a more dynamic format, like put in hours only when actually needed?

It will be harder and error-prone to parse...

**Comment by t-paul** [comment]
@MichaelAtOz what's that parsing about anyway? Easy parsing and user friendly output are sort-of opposites.

**Comment by Fat-Zer** [comment]
@t-paul, the issue suggests that this output might be used in a test suite and performance comparisons... It will require some basic text parsing of the log. It won't be hard to do as long as the output is pretty consistent, but by introducing of some disappearing value (e.g. flickering hour field) the date parsing would become quite painful, because you won't be able to just match it against a simple date format string e.g. `date` utility. Also if a user won't consider that the hour may appear on the output, he won't even bother to prepare the script for it.

The log-like output is pretty dumb. So it's a good fit as both: a user info messages and a parsable text stream for a specific information.

PS: sorry, for being over verbose, I haven't actually intended it ;)

**Comment by t-paul** [comment]
If we are talking test suite, I don't see much of an issue. Python would be able to parse this relatively easily and it's pretty unlikely there will be automatic testcases > 1h anyway. But in general it's a valid concern.

Another option would be to use the original format but allow fractional seconds.

**Comment by MichaelAtOz** [comment]
@t-paul yes as @Fat-Zer says, text parsing the log.

```
Total rendering time: h:mm:ss.zzz
```
Where `h` is just an int, so it would go to `hh` or `hhh` etc in those unlikely situations would be OK.

> are such renders anyway common? Or is it just a simple worst-case example?

24+ is not frequent*, but can happen, certainly >1h is not every day but happens, particularly with high $fn. If I was running it in the background I'd be unimpressed if it didn't actually show the time because it overflowed.

\* more often in the past when I didn't have 32GB and RAID SSDs

File: src/RenderStatistic.cc [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+{
+  GeometryCache::instance()->print();
+#ifdef ENABLE_CGAL
+  CGALCache::instance()->print();
+#endif
+}
+
+void RenderStatistic::printRenderingTime(std::chrono::milliseconds ms)
+{
+  PRINTB("Total rendering time: %02d:%02d.%02d",
```
Comment (t-paul) [inline src/RenderStatistic.cc]: Shouldn't the last one be `%03d`?

As discussed we also should add back the hours. I don't mind either way, just as leading value like `1:20:01.238` or going back to the original format (with added milliseconds).

**Comment by MichaelAtOz** [comment]
Almost...
The preview format should be the same format as render time. (below is GUI console)

>Parsing design (AST generation)...
Saved backup file: C:/Users/MeB/Documents/OpenSCAD/backups/unsaved-backup-llSPQEhJ.scad
Compiling design (CSG Tree generation)...
Compiling design (CSG Products generation)...
Geometries in cache: 146
Geometry cache size in bytes: 2559904
CGAL Polyhedrons in cache: 46
CGAL cache size in bytes: 67006192
Compiling design (CSG Products normalization)...
Normalized CSG tree has 1 elements
Compile and preview finished.
**Total rendering time: 0 hours, 0 minutes, 0 seconds**
>
>Parsing design (AST generation)...
Saved backup file: C:/Users/MeB/Documents/OpenSCAD/backups/unsaved-backup-llSPQEhJ.scad
Compiling design (CSG Tree generation)...
Rendering Polygon Mesh using CGAL...
Geometries in cache: 146
Geometry cache size in bytes: 2559904
CGAL Polyhedrons in cache: 46
CGAL cache size in bytes: 67006192
**Total rendering time: 00:00.21**
   Top level object is a 3D object:
   Facets:          6
Rendering finished.

As mentioned above, I'm happy with h:mm:ss.zzz, where h is flexible length.

and the preview command line output is not complete ie: 
e.g. `openscad -o cube.png cube.scad`

>Compiling design (CSG Products normalization)...
Normalized CSG tree has 1 elements

it should include the same cache info as the GUI preview console above.
@t-paul unless that code is not called from command line preview??

**Comment by kintel** [comment]
Minor request regarding the copyright notice: We request that authors allow us to adjust the license of this codebase in the future, either through copyright assignment, a more liberal copyright (like 2-clause BSD) on submitted code, or through an explicit grant like in https://github.com/openscad/openscad/blob/master/doc/contributor_copyright.txt.

**Comment by t-paul** [comment]
@MichaelAtOz right, good point. There's no statistics for preview (as there's no actual geometry in the general case). But the time output should match.

File: src/RenderStatistic.cc (hunk 1 above)
Comment (Fat-Zer) [reply src/RenderStatistic.cc]: thanks, good catch on the microseconds format.

**Comment by Fat-Zer** [comment]
Pushed suggested changes in a separate patch for review. Which includes:

 * change time format to `h:mm:ss.zzz` as suggested.
 * fix a time display for a preview in GUI.
 * print rendering statistic for rendering to PNG in preview mode
 * added missing virtual destructor declaration to `GLView` (mostly to suppress a warning)

If everything ok, I'll squash the patches...

@t-paul, fixed the time.

@MichaelAtOz, thanks, I've actually overlooked that preview prints anything... fixed both

PS: does it actually make any sense to print CGAL cache statistic for preview in both CLI and GUI? It supposed to be either zero or (for GUI) whatever left from the last run.

@kintel, I'm absolutely OK with granting you right to sublicense. But I'm not sure if I'm allowed to put a different license (or add a closure) to the file, because I've actually reused some lines from `mainwin.cc` which contains neither...

**Comment by t-paul** [comment]
I'd say we leave the cache part in as there's a feature request for persistent cache. If that gets implemented, the command line call would start showing cache content too. In addition there's a couple of modules that run though CGAL even in preview mode, so I would expect something to show up in case there's calls to `minkowski()`, `resize()` or one of the other modules forcing mesh calculation.

**Comment by MichaelAtOz** [comment]
- fix a time display for a preview in GUI.

Isn't working:
(OpenSCAD version 2020.03.09.ci4620)
>Compiling design (CSG Products generation)...
Geometries in cache: 124
Geometry cache size in bytes: 1907424
CGAL Polyhedrons in cache: 14
CGAL cache size in bytes: 10240096
Compiling design (CSG Products normalization)...
Normalized CSG tree has 72 elements
Compile and preview finished.
**Total rendering time: 0 hours, 0 minutes, 9 seconds**

Command line preview is good.

**Comment by Fat-Zer** [comment]
@MichaelAtOz, shouldn't be so... are you sure you rebuilt everything correctly/running correct version of the binary? if `openscad --version` displays a git hash as minor (`ci4620`, I'm not actually sure what it is) it should be `50cf23c3`...

By clicking on the preview in GUI I see:

    ...
    Total rendering time: 0:00:00.257

Are there any additional actions needed to reproduce that?

**Comment by MichaelAtOz** [comment]
> @MichaelAtOz, shouldn't be so... are you sure you running correct version of the binary?

@Fat-Zer oops sorry, I need to tidy up my Start menu...I didn't pin the new one as I usually do...
Capture-#3241

**Comment by t-paul** [comment]
@MichaelAtOz thanks for checking this out. I'll have another look, but I think we are good to go.

**Comment by t-paul** [comment]
@kintel anything open regarding your licensing question?

**Comment by Fat-Zer** [comment]
I'm squashed the fixup commit & rebased upon master

**Comment by kintel** [comment]
Added some minor cleanup and resolved a comment, plus the conflict here: #3330

**Comment by kintel** [comment]
I'm ok with the explicit right to sublicense given here.

**Pull request merged**